File: Algorithms/challenges/lc014_longest_common_prefix.py
"""
Time: O(n)
Space: O(1), excluding return.

Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".

Example 1:

Input: ["flower","flow","flight"]
Output: "fl"

Example 2:

Input: ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
Note:

All given inputs are in lowercase letters a-z.
"""


class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        cnt = 0
        for s in zip(*strs):
            if len(set(s)) == 1:  # or scan over s to avoid extra space for set
                cnt += 1
            else:
                break
        return strs[0][:cnt] if strs else ''


class Solution2(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        for i, letter_group in enumerate(zip(*strs)):
            if len(set(letter_group)) > 1:
                return strs[0][:i]
        return min(strs or [''])  # in case input is [''] or []

File: Algorithms/challenges/lc014_longest_common_prefix.cc
#include <string>
#include <vector>

using std::string;
using std::vector;


class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
      string prefix = "";
      for (unsigned i = 0; strs.size() > 0; prefix += strs[0][i], i++)
        for (unsigned j = 0; j != strs.size(); j++)
          if (i >= strs[j].size() || (j > 0 && strs[j][i] != strs[j-1][i]))
            return prefix;
      return prefix;
    }
};
